plugins/chat/app/services/auto_remove/handle_destroyed_group.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+module Chat
+  module Service
+    module AutoRemove
+      # Fired from [Jobs::AutoRemoveMembershipHandleUserRemovedFromGroup], which
+      # is in turn enqueued whenever the [DiscourseEvent] for :group_destroyed
+      # is triggered.
+      #
+      # The :group_destroyed event provides us with the user_ids of the former
+      # GroupUser records so we can scope this better.
+      #
+      # Since this could have potential wide-ranging impact, we have to check:
+      #   * The chat_allowed_groups [SiteSetting], and if any of the scoped users
+      #     are still allowed to use public chat channels based on this setting.
+      #   * The channel permissions of all the category chat channels the users
+      #     are a part of, based on [CategoryGroup] records
+      #
+      # If a user is in a groups that has the `full` or `create_post`
+      # [CategoryGroup#permission_types] or if the category has no groups remaining,
+      # then the user will remain in the channel.
+      class HandleDestroyedGroup
+        include Service::Base
+
+        contract
+        model :scoped_users
+        step :remove_users_outside_allowed_groups
+        step :remove_users_without_channel_permission
+        step :publish
+
+        class Contract
+          attribute :destroyed_group_user_ids
```

Comment:
Here also we should probably add some validation, otherwise this contract isn’t useful.